### DISCUSSION TITLE
Codebase feedback and questions (Max, March '23)

### DISCUSSION THREAD

**Discussion by maxdesalle** [sablier-labs/evm-monorepo#361] [Ideas]
Some minor suggestions and questions that came to mind while going through the contracts:
- Shouldn't we replace the ["V2" in the ascii art](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/SablierV2Comptroller.sol#L13) with "V2.0" given there will be new versions of the protocol?
- Is it normal that [the license for the SablierV2NftDescriptor.sol contract](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/SablierV2NftDescriptor.sol#L1) is unlicensed, as opposed to licensed with the BUSL-1.1? [Same for the interface contracts](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/interfaces/ISablierV2Lockup.sol#L1), they are all licensed under the GPL.
- Very minor thing, but it seems like there is [unnecessary spacing here](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/interfaces/ISablierV2Comptroller.sol#L62) (and in the below function descriptions too).
- Why is there [a limit on the fee](https://github.com/sablierhq/v2-core/blob/aaceec49eb9ee46a0d83b1308289032e8a28da22/src/interfaces/ISablierV2LockupLinear.sol#L104) frontends can charge?

**Comment by andreivladbrg** [comment]
> Shouldn't we replace the ["V2" in the ascii art](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/SablierV2Comptroller.sol#L13) with "V2.0" given there will be new versions of the protocol?

That's a good point. 

 > Is it normal that [the license for the SablierV2NftDescriptor.sol contract](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/SablierV2NftDescriptor.sol#L1) is unlicensed

That contract is just an [example](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/SablierV2NftDescriptor.sol#L8) used for scripts and tests. The actual implementation will be under BUSL.

> [Same for the interface contracts](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/interfaces/ISablierV2Lockup.sol#L1), they are all licensed under the GPL.

All the interfaces are left under the GPL because they are meant to be imported by our integrators. The actual implementation of them is under BUSL.

> it seems like there is [unnecessary spacing here](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/interfaces/ISablierV2Comptroller.sol#L62) 

The spacing between natspecs and comments has been intentionally left on all functions with longer documentation to improve readability.

> Why is there [a limit on the fee](https://github.com/sablierhq/v2-core/blob/aaceec49eb9ee46a0d83b1308289032e8a28da22/src/interfaces/ISablierV2LockupLinear.sol#L104) frontends can charge?

I will let @PaulRBerg answear this question.

**Comment by PaulRBerg** [comment]
Excellent feedback; thank you, @maxdesalle!

> Shouldn't we replace the ["V2" in the ascii art](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/SablierV2Comptroller.sol#L13) with "V2.0" given there will be new versions of the protocol?

It's an interesting observation. On the one hand, I would be inclined to say that we should keep "V2" in the ASCII art, since "V2" implicitly encompasses all versions (V2.0, V2.1, etc). On the other hand, using the dot notation would be more specific.

The catch with on-chain-specific versions is that we may only update some contracts on a subsequent new release. For instance, we could ship `SablierV2Comptroller` without touching any of the `SablierV2Lockup` contracts. Wouldn't it be somewhat confusing to have an on-chain system that represents our "V2.1" release, but which comprises contracts that contain "V2.0" in their ASCII art?

Let's sleep on this idea before making a decision.

> Is it normal that [the license for the SablierV2NftDescriptor.sol contract](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/SablierV2NftDescriptor.sol#L1) is unlicensed

@andreivladbrg answered this.

We will have the final implementation for the NFT descriptor only in April after we receive the visual assets from the graphic designers.

> [Same for the interface contracts](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/interfaces/ISablierV2Lockup.sol#L1), they are all licensed under the GPL

Again, @andreivladbrg answered this. There's more context on the licensing in the [README](https://github.com/sablierhq/v2-core/tree/d7f46c1212ebf9e56b93dfa1acec910b0193e0e9#licensing).

> there is [unnecessary spacing here](https://github.com/sablierhq/v2-core/blob/939319f3f4004ca56fb0e63d4acd3235ec2a174c/src/interfaces/ISablierV2Comptroller.sol#L62)

That spacing there is intentional. It serves as a visual delineator between the `@notice` and the `@dev` tags. The former is a simple explainer of what the function does, which is meant to be understood by everyone; the latter is a more advanced description aimed at developers.

> Why is there [a limit on the fee](https://github.com/sablierhq/v2-core/blob/aaceec49eb9ee46a0d83b1308289032e8a28da22/src/interfaces/ISablierV2LockupLinear.sol#L104) frontends can charge?

A host of reasons:

1. Gas optimization. It is easier to check for `MAX_FEE` (the same cap applied to protocol fees) than check for a bespoke 100% cap in the case of broker fees (note: the broker fee must not be greater than 100%, lest the ERC-20 transfer fails and other protocol assumptions break)
2. Security. Having a cap prevents accidental [decimal miscalculations](https://twitter.com/PaulRBerg/status/1634508717184909313).
3. A `MAX_FEE` of 10% should be enough for the vast majority of use cases (10% is what we believe `MAX_FEE` will be). If any hungry capitalist frontend wants to charge more than 10%, they could charge that high fee at their frontend level, and see the `broker.fee` to 0% in the create stream transaction.

**Reply by PaulRBerg** [reply]
I'd like to hear @razgraf's thoughts on the ASCII versioning issue, since this is, in part, a matter of product presentation.

**Reply by razgraf** [reply]
While referring to it as v2.0 is clearer, for the sake of simplicity I'd just keep V2 there (with the possibility of adding a smaller semver versioning number somewhere underneath). 

The split-deployment argument where only pieces are upgraded, not the entire protocol makes a good case for the "V2" naming.

**Reply by PaulRBerg** [reply]
Awesome, thanks for confirming.

A hardcoded semantic version would indeed be nice! Similar to the [one](https://github.com/PaulRBerg/prb-proxy/blob/4f647493e85dbd2c005c92df68cc47c5689d6d19/src/PRBProxyRegistry.sol#L16) I have in PRBProxy.

**Comment by PaulRBerg** [comment]
Closing, as we've settled and/ or responded to all points.
